### PR TITLE
[14.0] shopfloor_single_product_transfer: allow to scan packages

### PR DESCRIPTION
**Pull request by JuMiSanAr**
Now the user can scan packages on the first screen (before, only locations were allowed).

ref: rau-146

- [x] Add tests for scan_location endpoint.

Depends on:
- https://github.com/OCA/wms/pull/513
- https://github.com/OCA/wms/pull/520

### CODE REVIEW

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 1]
```diff
@@ -102,6 +108,34 @@ def _scan_location__check_stock(self, location):
             message = self.msg_store.location_empty(location)
             return self._response_for_select_location(message=message)
 
+    def _scan_location__check_package(self, location):
+        """Check if the location has lines without an assigned package."""
+        lines_without_package = self.env["stock.move.line"].search(
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: checking existing move lines should be only if allow_create_moves is False
In _scan_location__check_stock, I would also add a check that if all the quants in the location are part of a package, you have to scan a package

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 2]
```diff
@@ -102,6 +108,34 @@ def _scan_location__check_stock(self, location):
...
+            [
+                ("location_id", "=", location.id),
+                ("package_id", "=", False),
+                ("picking_id.picking_type_id", "in", self.picking_types.ids),
+            ]
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: add `"(state", "in", ("assigned", "partially_available"))`

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 3]
```diff
@@ -102,6 +108,34 @@ def _scan_location__check_stock(self, location):
...
+            return self._response_for_select_location(message=message)
+
+    def _scan_package__check_location(self, package):
+        """Check if this package corresponds to any of the allowed locations."""
+        locations = self.picking_types.default_location_src_id
+        child_locations = self.env["stock.location"].search(
+            [("id", "child_of", locations.ids)]
+        )
+        allowed_locations = locations | child_locations
+        if package.location_id not in allowed_locations:
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: Use instead `if not self.is_src_location_valid(package.location_id)`

Do the same change in `_scan_location__check_location`

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 4]
```diff
@@ -106,22 +106,21 @@ def _scan_location__check_stock(self, location):
...
-            return self._response_for_select_location(message=message)
+            return self._response_for_select_location_or_package(message=message)
 
     def _scan_location__check_package(self, location):
         """Check if the location has lines without an assigned package."""
         lines_without_package = self.env["stock.move.line"].search(
             [
                 ("location_id", "=", location.id),
                 ("package_id", "=", False),
-                ("picking_id.picking_type_id", "in", self.picking_types.ids),
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: I would have kept this `("picking_id.picking_type_id", "in", self.picking_types.ids)`

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 5]
```diff
@@ -325,29 +354,45 @@ def _use_handlers(self, handlers, *args, **kwargs):
             if response:
                 return response
 
+    def _add_location_package_lot_domain(
+        self, domain, location=None, package=None, lot=None
+    ):
+        if location:
+            domain = AND([domain, [("location_id", "=", location.id)]])
+        if package:
+            domain = AND([domain, [("package_id", "=", package.id)]])
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: If no package, do not match a line with package
`domain = AND([domain, [("package_id", "=", package.id if package else False)]])`

File: shopfloor_single_product_transfer/services/single_product_transfer.py [hunk 6]
```diff
@@ -356,11 +401,14 @@ def _create_move_from_location(
...
             "product_uom_qty": quantity,
-            "location_id": location.id,
             "location_dest_id": picking_type.default_location_dest_id.id,
             "origin": self.work.menu.name,
             "picking_type_id": picking_type.id,
         }
+        if location:
+            move_vals["location_id"] = location.id
+        if package:
+            move_vals["package_id"] = package.id
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer/services/single_product_transfer.py]: This field does not exist. You need to leverage forced_package_id in the action_assign().
I would create a package_level and set it on the move after action_confirm. This way the move has a picking_id you can use on the package level creation.

**Review by jbaudoux — CHANGES_REQUESTED** [review]

**Comment by JuMiSanAr** [comment]
Closed as this work is already included in https://github.com/OCA/wms/pull/640

**Pull request closed without merging**